### ISSUE TITLE
Allow specifying injectivity during promotion

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#243]
This is an injective function:

```haskell
not :: Bool -> Bool
not False = True
not True  = False
```

Its promoted counterpart, however:

```haskell
type family Not (a :: Bool) :: Bool where
  Not False = True
  Not True  = False
```

is not explicitly marked as injective in `a`, so GHC can't take advantage of that knowledge. It would be nifty if one could indicate to `singletons` that it should try to promote a particular function using certain type family dependencies. The challenging part would be coming up with a good API for this.

**Comment by mstksg** [comment]
Maybe, alongside

```
genPromotions :: DsMonad q => [Name] -> q [Dec]
```

We could have a `genPromotionsInjective` ?

**Comment by RyanGlScott** [comment]
It'd need to be more involved than that. The issue is that there are some functions which are injective in some (but not all) of its arguments. For instance:

```haskell
const :: a -> b -> a
const x _ = x
```

`const` is injective in its first argument, but not its second, so we'd need a more to inform `singletons` that it should be promoted as:

```haskell
type family Const a b = r | r -> a where
  Const a _ = a
```

**Comment by goldfirere** [comment]
Isn't this what TH annotations are for? Look up the details of the `{-# ANN ... #-}` pragma, which (I believe) allows you to associate extra info with a declaration.

**Comment by RyanGlScott** [comment]
Sure, annotations would be one way of communicating this information to Template Haskell. But the thing I'm most unsure of is an API for _encoding_ which arguments should be marked as injective in an annotation (or whatever mechanism we decide one).

**Comment by RyanGlScott** [comment]
At the moment, the best I can think of is a pair:

```
(String, [Int])
```

where the `String` represents the name of the thing to be promoted (we can't use a `Name`, since it might be a not-yet-spliced-in function name), and the `[Int]` represent the indices of the arguments which should be marked as injective. Of course, this encoding has its flaws, since an index might be out of range, but I can't way to statically encode this invariant, so I suppose we could just sanity-check that every index is in range during promotion.

**Comment by goldfirere** [comment]
You don't need the string, as annotation are always attached to declarations (in this case, the declaration of the function).

Indexing is ugly, but it gets the job done, and it will be the tersest API. The other thing that's possible is a string with parameter names and arrows that has to be parsed.

**Comment by RyanGlScott** [comment]
Indexing sounds like the way to go then.

However, I'm skeptical that we'll be able to make this work via annotations. The reason being is that we'd need to account for locally defined annotations like:

```haskell
$(promote [d| not :: Bool -> Bool
              not False = True
              not True  = False
              {-# ANN not (InjIndices [0]) #-} |])
```

However, Template Haskell doesn't provide a very satisfying interface for implementing `reifyAnnotations`, since all the information it gives you about annotation pragmas is [simply](http://hackage.haskell.org/package/template-haskell-2.12.0.0/docs/Language-Haskell-TH-Syntax.html#t:Pragma):

```haskell
data Pragma
  = ...
  | AnnP AnnTarget Exp
```

And I fail to see how you'd implement `qReifyAnnotations :: (Quasi q, Data a) => AnnLookup -> q [a]` using just an `Exp`.

**Comment by goldfirere** [comment]
The `Exp` can be whatever you like. For instance, it could be a list of indices, a `ListE [LitE (IntegerL 2), LitE (IntegerL 4)]` or some such.

**Comment by RyanGlScott** [comment]
> The Exp can be whatever you like.

But therein lies the rub. There isn't a way to take an `Exp` and coerce it into the API that Template Haskell offers for reifying annotations, which is `qReifyAnnotations :: (Quasi q, Data a) => AnnLookup -> q [a]`. That is to say, one has to be able to take an arbitrary `Exp` and fit it into the `Data` API. AFAICT, there is no way to accomplish that.

**Comment by goldfirere** [comment]
Aha! I see the problem. You're putting the `ANN` into the quote, and I was thinking of it outside the quote. Actually, either one would work.

Inside the quote: You would indeed have an `AnnP`. But all you need is a pass through the declarations looking for injectivity annotations, extract the information from the `Exp`, and then use that information during promotion. No need for `reifyAnnotations` here.

Outside the quote: Just call `reifyAnnotations` expecting a `[Int]`.

Or am I missing something?

**Comment by RyanGlScott** [comment]
> Inside the quote: You would indeed have an `AnnP`. But all you need is a pass through the declarations looking for injectivity annotations, extract the information from the `Exp`, and then use that information during promotion. No need for `reifyAnnotations` here.

I don't think it's that simple. After all, you can put arbitrary expressions into `ANN`s:

```haskell
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}

import Language.Haskell.TH
import System.Info

f :: Int
f = 42

{-# ANN f os #-}

$(return [])

main :: IO ()
main = putStrLn $(reifyAnnotations (AnnLookupName 'f) >>= \(anns :: [String]) -> stringE (show anns))
```

```
$ runghc Main.hs
["linux"]
```

However, if `{-# ANN f os #-}` were quoted, all the info we'd have would be `AnnP (ValueAnnotation 'f) (VarE 'os)`. It's not clear to me how we'd go from `VarE 'os` to `"linux"` with just the `template-haskell` API.

**Comment by goldfirere** [comment]
So you're worried about a programmer, say, setting `indices :: [Int]; indices = [1,2,4]` and then using `indices` in their annotations? Just say they can't. :) All annotation index lists must be literal. I still feel like I'm missing something here...

**Comment by RyanGlScott** [comment]
Well, that is one option. A consequence of that design choice is that the expressiveness of quoted `ANN`s would be necessarily much more limited than non-quoted `ANN`s. But I don't really see a better API for this, so I suppose I can live with that.

**Comment by goldfirere** [comment]
Now I'm confused again. I don't think we need to support both quoted and unquoted `ANN`s. We just say which one is the API for this feature. Do whichever is more convenient.

**Comment by RyanGlScott** [comment]
I claim that we would need to support quoted `ANN`s. After all, there's a serious staging issue that presents itself. You can't put an unquoted `ANN` declaration before the quoted things are spliced in:

```
{-# LANGUAGE TemplateHaskell #-}
module Foo where

{-# ANN f "test" #-}

$([d| f :: Int
      f = 42 |])
```
```
Foo.hs:4:1: error:
    • Not in scope: ‘f’
    • In the annotation: {-# ANN f "test" #-}
  |
4 | {-# ANN f "test" #-}
  | ^^^^^^^^^^^^^^^^^^^^
```

But you also can't put an unquoted `ANN` after the quoted things are spliced in either. That is, while this compiles:

```
{-# LANGUAGE TemplateHaskell #-}
module Foo where

$([d| f :: Int
      f = 42 |])

{-# ANN f "test" #-}
```

There's a problem here in that the annotation for `f` needs to be present at the time of splicing. However, if you try to reify what `f`'s annotations are there:

```
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
module Foo where

import Language.Haskell.TH

$(do decs <- [d| f :: Int
                 f = 42 |]
     anns <- reifyAnnotations (AnnLookupName (mkName "f"))
               >>= \(anns :: [String]) -> stringE (show anns)
     runIO (print anns)
     return decs)

{-# ANN f "test" #-}
```

Then it'll fail, unsurprisingly:

```
Foo.hs:7:3: error: ‘f’ is not in scope at a reify
  |
7 | $(do decs <- [d| f :: Int
  |   ^^^^^^^^^^^^^^^^^^^^^^^...
```

Therefore, in order to reify annotations for quoted things, you really do need to use quoted `ANN`s (and similarly for unquoted things). There is no one-size-fits-all solution, and unfortunately the two mechanisms we'd need to support to make this work (`reifyAnnotations` for unquoted `ANN`s, and our home-grown solution for quoted `ANN`s) have different expressive capabilities. That's what worries me.

**Comment by goldfirere** [comment]
You've convinced me that quoted `ANN`s are necessary. What about unquoted ones? Do we ever promote something unquoted? I suppose we do, in, say, classes. Perhaps you're right that we need both. That's a bit maddening.

What if we avoid using `ANN` directly but be inspired by it. We could have a function `injectivity :: String -> [Int] -> Q [Dec]` that takes the name of a function, its injectivity descriptor, and always returns an empty list of decls. This would store information somewhere about the injectivity of the function with the given name. Note that the name does not have to be in scope yet. Then, when promoting, we just retrieve injectivity information as we need it.

Where to store the information? In the `Q` monad, using `putQ` and `getQ`, which could work with a `Map String [Int]` or some such. (At first blush, I thought we could promote the user-supplied string to a `Symbol` and then use the `Typeable`-based lookup used in `getQ` and `putQ`, but I actually don't think this is possible, unfortunately. Someone needs to get around to implementing dependent types.) There may be issues around naming (qualified vs. unqualified, ambiguity) because we won't be using GHC's renamer, but we can figure out how to deal with these.

How does this plan sound?

**Comment by RyanGlScott** [comment]
I like the sound of this much better. Using a top-level function to mutate the `Q` state somewhat constrains the way that users can specify injectivity (in the sense that they must call `injectivity` _before_ `promote`), but this has the fortunate side effect that we don't need to worry about the quoted/non-quoted distinction at all.

> There may be issues around naming (qualified vs. unqualified, ambiguity) because we won't be using GHC's renamer, but we can figure out how to deal with these.

True. But for better or worse, the design of `singletons` itself obviates many of these issues. For instance, we don't really have to worry about name ambiguity too much, since the only things we care about specifying injectivity for are top-level function names. If we give an ambiguous name to `injectivity`, then it seems almost guaranteed that the corresponding function would also be rejected with an ambiguity error, so we're not making things worse with this design.

Also, `Q`'s state is only maintained on a per-module basis, so we don't need to worry about having the injectivity of a function in one module clobber an identical name in a separate module.

**Comment by RyanGlScott** [comment]
Oh dear. When attempting to sketch out an implementation of this, I realized that there is an _enormous_ problem in that the way `singletons` promotes things effectively prevents us from specifying the injectivity properties for most promoted type families. Here is a small example:

```haskell
const :: a -> b -> a
const x _ = x

asTypeOf :: a -> a -> a
asTypeOf = const
```

When promoted, these become:

```haskell
type family Const (x :: a) (y :: b) :: a where
  Const x _ = x

type family AsTypeOf (x :: a) (y :: b) :: a where
  AsTypeOf a b = ConstSym0 @@ a @@ b
```

Both of these functions are injective in their first argument. Specifying this for `Const` goes through with no problems:

```haskell
type family Const (x :: a) (y :: b) = (r :: a) | r -> a where
  Const x _ = x
```

But doing so for `AsTypeOf`:

```haskell
type family AsTypeOf (x :: a) (y :: b) = (r :: a) | r -> a where
   AsTypeOf a b = ConstSym0 @@ a @@ b
```

Gets rejected by GHC:

```
<interactive>:19:66: error:
    • Type family equation violates injectivity annotation.
      RHS of injective type family equation cannot be a type family:
        forall k1 k (a :: k) (b :: k1).
          AsTypeOf a b = (ConstSym0 @@ a) @@ b
            -- Defined at <interactive>:19:66
    • In the equations for closed type family ‘AsTypeOf’
      In the type family declaration for ‘AsTypeOf’
```

In other words, the use of `(@@)` stops us in our tracks. This is an issue for any definition that uses function application—which, as it turns out, is most of the functions in `singletons`.

In fact, it's not just function application that's the culprit—`case` statements can also throw a wrench into injectivity. For instance, the `absurd` function from `Data.Void`:

```haskell
absurd :: Void -> a
absurd x = case x of {}
```

Could conceivably be promoted to a form which is injective:

```haskell
type family Absurd (x :: Void) = (r :: a) | r -> x
```

But in practice, it's actually promoted to something like this:

```haskell
type family Absurd (x :: Void) :: a where
  Absurd x = Case_12345 x

type family Case_12345 (x :: Void) :: a
```

Alas, the use of the `Case_12345` type family on the head of the RHS prevents `Absurd` from being injective for the same reasons as before.

-----

Unlike in other `singletons` bugs, where we're blocked by some GHC bug, I think GHC is quite right to reject these definitions as non-injective. After all, injectivity is not a compositional property, as the [Injective Type Families](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/07/injective-type-families-acm.pdf?from=https%3A%2F%2Fresearch.microsoft.com%2Fen-us%2Fum%2Fpeople%2Fsimonpj%2Fpapers%2Fext-f%2Finjective-type-families-acm.pdf) paper lays out in section 4.1. This leads me to believe that this goal is too ambitious.

**Comment by RyanGlScott** [comment]
It's also worth noting that there isn't exactly a wealth of functions in `singletons` that can be marked injective to begin with. I did a quick scan of the `Data.Promotion.*` module namespace to find things that could be conceivably be marked as injective, and these were the only candidates I found:

* `Data.Promotion.Prelude.Base`
  - `Id`
  - `Const`
  - `AsTypeOf`
  - `Seq`
* `Data.Promotion.Prelude.Bool`
  - `Not`
* `Data.Promotion.Prelude.List`
  - `Reverse`
* `Data.Promotion.Prelude.Tuple`
  - `Swap`
  - `Unzip{3,4,5,6}`
* `Data.Promotion.Prelude.Void`
  - `Absurd`

If we want to make these injective, then the simplest solution might be to simply hard-code their promoted definitions, as unappealing as that sounds...

**Comment by goldfirere** [comment]
I agree completely with these recent pessimistic comments. If we had a thoroughly first-class notion of injectivity in GHC, then we could imagine somehow tagging defunctionalization symbols with injectivity info and then designing `(@@)` to use these... and doing much more magic besides. But we don't have first-class injectivity in GHC.

I do think this observation kills the idea of injectivity via singletons, sad to say.
